### PR TITLE
fix: Element access operator should derive index type properly from overloaded signature

### PR DESCRIPTION
**Pull request by MaxGraey**
- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/compiler.ts [hunk 1]
```diff
@@ -9164,6 +9173,7 @@ export class Compiler extends DiagnosticEmitter {
         this.currentType,
         expression.operand,
         resolver.currentThisExpression,
+        Type.i32,
```
Comment (MaxGraey) [inline src/compiler.ts:9176]: @dcodeIO It looks like we should use proper index type also for `compileUnaryPostfixExpression` and `compileUnaryPrefixExpression`. Could you give me a hint as to the best way to derive this in these methods?

File: src/compiler.ts [hunk 2]
```diff
@@ -5735,7 +5738,17 @@ export class Compiler extends DiagnosticEmitter {
...
           Constraints.CONV_IMPLICIT | Constraints.IS_THIS
         );
-        let elementExpr = this.compileExpression(assert(indexExpression), Type.i32, Constraints.CONV_IMPLICIT);
+        let setterIndexType = setterInstance.signature.parameterTypes[0];
+        let getterIndexType = getterInstance.signature.parameterTypes[0];
+        if (!setterIndexType.equals(getterIndexType)) {
+          this.error(
+            DiagnosticCode.Index_signature_type_should_be_the_same_for_element_accessors_in_type_0,
+            valueExpression.range, classInstance.internalName
+          );
```
Comment (dcodeIO) [inline src/compiler.ts:5747]: The closest there is in TS is probably

```
The return type of a 'get' accessor must be assignable to its 'set' accessor type(2380)
```

Would that work here?

Comment (MaxGraey) [reply src/compiler.ts:5747]: Not sure. Overloaded element accessors is little bit more special than simple `get` and `set`.

Comment (dcodeIO) [inline src/compiler.ts:5747]: Can this perhaps use `errorRelated` to show both the signature of the setter and getter? :)

File: src/diagnosticMessages.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
   "Start function name '{0}' is invalid or conflicts with another export.": 110,
   "Element '{0}' not found.": 111,
   "Exchange of '{0}' values is not supported by all embeddings": 112,
+  "Index signature type should be the same for element accessors in type '{0}'.": 113,
```
Comment (dcodeIO) [inline src/diagnosticMessages.json:15]:
```suggestion
  "Index signature in type '{0}' differs in types.": 113,
```
...then with errorRelated showing both places? :)

File: src/compiler.ts [hunk 4]
```diff
@@ -5735,7 +5738,19 @@ export class Compiler extends DiagnosticEmitter {
...
-        let elementExpr = this.compileExpression(assert(indexExpression), Type.i32, Constraints.CONV_IMPLICIT);
+        let setterIndexType = setterInstance.signature.parameterTypes[0];
+        let getterIndexType = getterInstance.signature.parameterTypes[0];
+        if (!setterIndexType.equals(getterIndexType)) {
+          this.errorRelated(
+            DiagnosticCode.Index_signature_accessors_in_type_0_differ_in_types,
+            getterInstance.identifierAndSignatureRange,
+            setterInstance.identifierAndSignatureRange,
+            classInstance.internalName,
+          );
```
Comment (MaxGraey) [inline src/compiler.ts:5749]: @dcodeIO Is it right usage of `errorRelated`?

**Comment by JairusSW** [comment]
Thank you! :D

**Pull request merged**